File: android/app/src/main/kotlin/in/nann/smashhockey/scene/Actors.kt
package `in`.nann.smashhockey.scene

import android.opengl.Matrix
import com.google.android.filament.Engine
import com.google.android.filament.MaterialInstance
import com.google.android.filament.Scene
import `in`.nann.smashhockey.core.generated.Role
import `in`.nann.smashhockey.core.generated.Sport
import `in`.nann.smashhockey.core.generated.Tuning
import `in`.nann.smashhockey.core.match.MatchSnapshot
import `in`.nann.smashhockey.engine.Geometry
import `in`.nann.smashhockey.engine.Materials
import `in`.nann.smashhockey.engine.MeshKit
import `in`.nann.smashhockey.engine.Node
import `in`.nann.smashhockey.generated.Presentation
import `in`.nann.smashhockey.generated.WorldLook
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

/** A side's kit colours, sRGB 0xRRGGBB. */
data class TeamColours(val primary: Int, val secondary: Int)

/**
 * Everything that moves in a match, drawn from its snapshot (spec §5, §8): the twelve players as the
 * prototype's disks (ADR 0006), the ball or puck — rolling or spinning, with its trail (§8.8) — the
 * aim arrow (§5.2, [AimArrow]) and the ring under a player the whistle would name offside (§8.9,
 * §16.4); the twin of iOS's Actors.swift.
 *
 * Between ticks it extrapolates by at most one tick (`ahead`, match seconds) along the snapshot's
 * velocities — drawing only; the simulation is advanced by the core's tick clock alone (§4.2).
 */
class Actors(
    private val engine: Engine,
    private val scene: Scene,
    parent: Int,
    first: MatchSnapshot,
    colours: List<TeamColours>,
    private val sport: Sport,
    orbitPeriod: Double,
    materials: Materials,
    look: WorldLook,
) {
    private val p = Presentation.Player
    private val omega = 2 * PI / orbitPeriod
    private val geometries = HashMap<String, Geometry>()
    private val entities = ArrayList<Int>()
    private val players: List<Node>
    private class Shadow(val entity: Int, val rest: MaterialInstance, val carrier: MaterialInstance)
    private val shadows = ArrayList<Shadow>()
    private var carrierShown: Int? = null
    private val ball: Node
    private val ballDisc: Node
    private val ballRadius = first.ball.radius
    private val aim = AimArrow(engine, scene, parent, materials, sport.cornerRadius)
    private val trail = Trail(engine, scene, materials, first.ball.radius.toFloat())
    private val offside = OffsideMarks(engine, scene, parent, first, colours, sport, materials)
    /** The ball's accumulated roll (field) or spin (puck), and where it was last drawn. */
    private val spin = FloatArray(16).also { Matrix.setIdentityM(it, 0) }
    private var lastBall: DoubleArray? = null
    private var lastClock = 0.0

    init {
        fun geo(key: String, kit: () -> MeshKit) = geometries.getOrPut(key) { Geometry(engine, kit()) }
        /** A renderable of [g] in [m] under [under]; flat marks draw after what they lie on. */
        fun part(g: Geometry, m: MaterialInstance, under: Int, priority: Int = 4): Node {
            val e = g.renderable(listOf(m), shadows = false, priority = priority)
            scene.addEntity(e); entities += e
            return Node(engine, under, existing = e)
        }
        val d = Presentation.Dummy
        players = first.players.map { pl ->
            val r = pl.radius.toFloat()
            val dummy = pl.role == Role.DUMMY
            val primary = if (dummy) d.body else colours[pl.team].primary
            val secondary = if (dummy) d.stripe else colours[pl.team].secondary
            val h = (if (dummy) d.height else p.height).toFloat()
            val group = Node(engine, parent)
            part(geo("body$r/$h") { Shapes.body(r, h) }, materials.toon(primary, look), group.entity)
            when (pl.role) {
                Role.GOALIE -> part(geo("ring$r") { Shapes.goalieRing(r, h) }, materials.flat(secondary), group.entity, 5)
                Role.DUMMY -> part(geo("stripe$r") { Shapes.stripe(r) }, materials.toon(secondary, look), group.entity)
                else -> part(geo("dot$r") { Shapes.dot(r, h) }, materials.flat(secondary), group.entity, 5)
            }
            val rest = materials.flat(primary, p.shadowOpacity)
            val shadow = part(geo("shadow$r") { Shapes.shadow(r) }, rest, group.entity, 5)
            shadows += Shadow(shadow.entity, rest, materials.flat(primary, p.carrierShadowOpacity))
            group
        }

        val b = Presentation.Ball
        val r = first.ball.radius.toFloat()
        ball = if (sport == Sport.ICE) part(geo("puck") { Shapes.puck(r, b.puckHeight.toFloat()) }, materials.toon(b.ice, look), parent)
        else part(geo("ball") { Shapes.ball(r) }, materials.toon(b.field, look), parent)
        ballDisc = part(geo("disc") { Shapes.disc() }, materials.flat(b.disc, b.discOpacity), parent, 5)
        ballDisc.scale(b.discRadius.toFloat())
    }

    /** Draws snapshot [s], [ahead] match seconds past its tick; [celebrating]'s players hop on real [clock]. */
    fun update(s: MatchSnapshot, ahead: Double, celebrating: Int?, clock: Double) {
        val dt = clock - lastClock              // taken before `roll` moves `lastClock` on
        val xs = DoubleArray(s.players.size); val zs = DoubleArray(s.players.size)
        s.players.forEachIndexed { i, pl ->
            xs[i] = pl.x + pl.vx * ahead; zs[i] = pl.z + pl.vz * ahead
            val n = players[i]
            n.x = xs[i].toFloat(); n.z = zs[i].toFloat()
            n.y = if (celebrating == pl.team && pl.role != Role.DUMMY) (p.hop * abs(sin(clock * p.hopRate + i * 0.9))).toFloat() else 0f
            // A running player leans into its run, as the prototype's did.
            val speed = sqrt(pl.vx * pl.vx + pl.vz * pl.vz)
            val lean = if (speed > 0.5) min(speed * p.lean, p.maxLean) else 0.0
            n.pitch = if (speed > 0.5) (pl.vz / speed * lean).toFloat() else 0f
            n.roll = if (speed > 0.5) (-pl.vx / speed * lean).toFloat() else 0f
            n.apply()
        }

        // The disc under the carrier darkens.
        val carrier = s.ball.carrier
        if (carrier != carrierShown) {
            val rm = engine.renderableManager
            carrierShown?.let { rm.setMaterialInstanceAt(rm.getInstance(shadows[it].entity), 0, shadows[it].rest) }
            carrier?.let { rm.setMaterialInstanceAt(rm.getInstance(shadows[it].entity), 0, shadows[it].carrier) }
            carrierShown = carrier
        }

        val b = s.ball
        var bx = b.x + b.vx * ahead; var bz = b.z + b.vz * ahead
        var angle = b.orbit
        val c = b.carrier
        if (c != null) {
            angle += b.orbitDirection * omega * ahead
            bx = xs[c] + Tuning.Orbit.radius * sin(angle); bz = zs[c] + Tuning.Orbit.radius * cos(angle)
        }
        roll(bx, bz, clock)
        ball.x = bx.toFloat(); ball.y = if (sport == Sport.ICE) 0f else ballRadius.toFloat(); ball.z = bz.toFloat(); ball.apply()
        ballDisc.x = bx.toFloat(); ballDisc.y = 0.02f; ballDisc.z = bz.toFloat(); ballDisc.apply()
        trail.update(bx, bz, s.time + ahead, sqrt(b.vx * b.vx + b.vz * b.vz), c == null)
        aim.update(s, xs, zs, angle, clock)
        offside.update(s, xs, zs, dt)
    }

    /**
     * §8.8: the field ball rolls by its drawn travel — |Δ| / r about (Δz, 0, −Δx) — the puck spins
     * about its axis at a steady rate of real time.
     */
    private fun roll(x: Double, z: Double, clock: Double) {
        val dt = clock - lastClock
        lastClock = clock
        val last = lastBall
        lastBall = doubleArrayOf(x, z)
        val step = FloatArray(16)
        if (sport == Sport.ICE) {
            Matrix.setRotateM(step, 0, Math.toDegrees(Presentation.Trail.puckSpin * dt).toFloat(), 0f, 1f, 0f)
        } else {
            if (last == null) return
            val dx = x - last[0]; val dz = z - last[1]
            val d = sqrt(dx * dx + dz * dz)
            if (d < 1e-6 || d > 5) return          // a restart teleports the ball: no spin for that
            Matrix.setRotateM(step, 0, Math.toDegrees(d / ballRadius).toFloat(), (dz / d).toFloat(), 0f, (-dx / d).toFloat())
        }
        val out = FloatArray(16)
        Matrix.multiplyMM(out, 0, step, 0, spin, 0)
        System.arraycopy(out, 0, spin, 0, 16)
        ball.rotation = spin
    }

    fun destroy() {
        aim.destroy()
        trail.destroy()
        offside.destroy()
        entities.forEach { scene.removeEntity(it) }
        geometries.values.forEach { it.destroy() }
        players.forEach { engine.destroyEntity(it.entity); com.google.android.filament.EntityManager.get().destroy(it.entity) }
    }
}

File: ios/Sources/Scene/Actors.swift
import RealityKit
import SmashCore
import simd

/// A side's kit colours, sRGB 0xRRGGBB.
struct TeamColours: Equatable {
    var primary: UInt32
    var secondary: UInt32
}

/// Everything that moves in a match, drawn from its snapshot (spec §5, §8): the twelve players as
/// the prototype's disks (ADR 0006), the ball or puck, the orbit ring and the aim line (§5.2). The
/// twin of Android's Actors.kt.
///
/// Between ticks it extrapolates by at most one tick (`ahead`, match seconds) along the snapshot's
/// velocities — drawing only; the simulation is advanced by the core's tick clock alone (§4.2).
@MainActor
final class Actors {
    typealias P = Presentation.Player
    let root = Entity()
    private var players: [Entity] = []
    private var shadows: [(entity: ModelEntity, rest: UnlitMaterial, carrier: UnlitMaterial)] = []
    private var carrierShown: Int?
    private let ball: ModelEntity
    private let ballDisc: ModelEntity
    private let orbit: ModelEntity
    private let aim: ModelEntity
    private let arrow: ModelEntity
    private let target: ModelEntity
    private let aimColours: [UnlitMaterial]     // pass, shot, free
    private var aimKind = -1
    private let omega: Double

    init(first s: MatchSnapshot, colours: [TeamColours], sport: Sport, orbitPeriod: Double,
         materials: Materials, look: WorldLook) throws {
        omega = 2 * .pi / orbitPeriod
        typealias D = Presentation.Dummy
        var meshes: [String: MeshResource] = [:]
        func mesh(_ key: String, _ kit: @autoclosure () -> MeshKit) throws -> MeshResource {
            if let m = meshes[key] { return m }
            let m = try kit().resource()
            meshes[key] = m
            return m
        }
        func model(_ mesh: MeshResource, _ material: RealityKit.Material) -> ModelEntity {
            ModelEntity(mesh: mesh, materials: [material])
        }

        for p in s.players {
            let r = Float(p.radius)
            let group = Entity()
            let dummy = p.role == .dummy
            let primary = dummy ? D.body : colours[p.team].primary
            let secondary = dummy ? D.stripe : colours[p.team].secondary
            let h = Float(dummy ? D.height : P.height)
            group.addChild(model(try mesh("body\(r)/\(h)", Shapes.body(radius: r, height: h)),
                                 try materials.toon(primary, look: look)))
            switch p.role {
            case .goalie:
                group.addChild(model(try mesh("ring\(r)", Shapes.goalieRing(radius: r, height: h)), Materials.flat(secondary)))
            case .dummy:
                group.addChild(model(try mesh("stripe\(r)", Shapes.stripe(radius: r)), try materials.toon(secondary, look: look)))
            case .defender, .forward:
                group.addChild(model(try mesh("dot\(r)", Shapes.dot(radius: r, height: h)), Materials.flat(secondary)))
            }
            let rest = Materials.flat(primary, opacity: P.shadowOpacity)
            let shadow = model(try mesh("shadow\(r)", Shapes.shadow(radius: r)), rest)
            group.addChild(shadow)
            shadows.append((shadow, rest, Materials.flat(primary, opacity: P.carrierShadowOpacity)))
            root.addChild(group)
            players.append(group)
        }

        typealias B = Presentation.Ball
        let r = Float(s.ball.radius)
        if sport == .ice {
            ball = model(try Shapes.puck(radius: r, height: Float(B.puckHeight)).resource(), try materials.toon(B.ice, look: look))
        } else {
            ball = model(try Shapes.ball(radius: r).resource(), try materials.toon(B.field, look: look))
        }
        ballDisc = model(try Shapes.disc().resource(), Materials.flat(B.disc, opacity: B.discOpacity))
        ballDisc.scale = SIMD3(repeating: Float(B.discRadius))
        root.addChild(ballDisc)
        root.addChild(ball)

        typealias A = Presentation.Aim
        func mark(_ kit: MeshKit, _ m: UnlitMaterial) throws -> ModelEntity {
            let e = ModelEntity(mesh: try kit.resource(), materials: [m])
            e.isEnabled = false
            return e
        }
        let orbitRadius = Float(Tuning.Orbit.radius), half = Float(A.orbitWidth) / 2
        orbit = try mark(Shapes.ring(inner: orbitRadius - half, outer: orbitRadius + half, segments: 48),
                         Materials.flat(A.orbit, opacity: A.orbitOpacity))
        aimColours = [A.pass, A.shot, A.free].map { Materials.flat($0, opacity: A.opacity) }
        aim = try mark(Shapes.strip(), aimColours[0])
        arrow = try mark(Shapes.arrowhead(), aimColours[0])
        let rr = Float(Tuning.Player.outfieldRadius)
        target = try mark(Shapes.ring(inner: rr + Float(P.targetInner), outer: rr + Float(P.targetOuter), segments: 32),
                          Materials.flat(P.target, opacity: P.targetOpacity))
        for e in [orbit, aim, arrow, target] { root.addChild(e) }
    }

    /// Draws snapshot `s`, `ahead` match seconds past its tick (0 ≤ ahead < one tick).
    /// `celebrating` is the team that just scored, whose players hop on real time `clock`.
    func update(_ s: MatchSnapshot, ahead: Double, celebrating: Int?, clock: Double) {
        var positions: [SIMD2<Double>] = []
        for (i, p) in s.players.enumerated() {
            let pos = SIMD2(p.x + p.vx * ahead, p.z + p.vz * ahead)
            positions.append(pos)
            var y: Float = 0
            if let team = celebrating, team == p.team, p.role != .dummy {
                y = Float(P.hop * abs(sin(clock * P.hopRate + Double(i) * 0.9)))
            }
            let e = players[i]
            e.position = SIMD3(Float(pos.x), y, Float(pos.y))
            // A running player leans into its run, as the prototype's did.
            let speed = (p.vx * p.vx + p.vz * p.vz).squareRoot()
            let lean = speed > 0.5 ? min(speed * P.lean, P.maxLean) : 0
            let pitch = speed > 0.5 ? Float(p.vz / speed * lean) : 0, roll = speed > 0.5 ? Float(-p.vx / speed * lean) : 0
            e.orientation = simd_quatf(angle: pitch, axis: [1, 0, 0]) * simd_quatf(angle: roll, axis: [0, 0, 1])
        }

        // The disc under the carrier darkens.
        let carrier = s.ball.carrier
        if carrier != carrierShown {
            if let old = carrierShown { shadows[old].entity.model?.materials = [shadows[old].rest] }
            if let c = carrier { shadows[c].entity.model?.materials = [shadows[c].carrier] }
            carrierShown = carrier
        }

        // The ball: on its orbit round the carrier (§5.1), else where it rolls.
        let b = s.ball
        var ballPos = SIMD2(b.x + b.vx * ahead, b.z + b.vz * ahead)
        var angle = b.orbit
        if let c = b.carrier {
            angle += b.orbitDirection * omega * ahead
            ballPos = positions[c] + Tuning.Orbit.radius * SIMD2(sin(angle), cos(angle))
            orbit.position = SIMD3(Float(positions[c].x), 0.025, Float(positions[c].y))
        }
        ball.position = SIMD3(Float(ballPos.x), 0, Float(ballPos.y))
        orbit.isEnabled = b.carrier != nil
        ballDisc.position = SIMD3(Float(ballPos.x), 0.02, Float(ballPos.y))
        drawAim(s, positions: positions, ball: ballPos, angle: angle, clock: clock)
    }

    /// The aim line (§5.2): from the ball toward where a release now would go, coloured by what it
    /// would snap to — a pass (with the green ring under the receiver), a shot, or nothing.
    private func drawAim(_ s: MatchSnapshot, positions: [SIMD2<Double>], ball: SIMD2<Double>, angle: Double, clock: Double) {
        typealias A = Presentation.Aim
        guard s.playerCarrier, let c = s.ball.carrier, let kind = s.aim, s.state == .play || s.state == .ready else {
            aim.isEnabled = false; arrow.isEnabled = false; target.isEnabled = false
            return
        }
        let me = positions[c]
        var end: SIMD2<Double>
        var colour = 2
        target.isEnabled = false
        switch kind {
        case .pass(let m):
            let p = s.players[m]
            let d = simd_distance(positions[m], me)
            let t = d / max(Tuning.Orbit.leadSpeedMin, Tuning.Orbit.leadSpeedBase + Tuning.Orbit.leadSpeedPerMetre * d)
            let lead = positions[m] + Tuning.Orbit.leadVelocityFactor * SIMD2(p.vx, p.vz) * t
            end = ball + simd_normalize(lead - me) * simd_distance(lead, ball)
            colour = 0
            target.isEnabled = true
            target.position = SIMD3(Float(positions[m].x), 0.03, Float(positions[m].y))
            target.scale = SIMD3(repeating: Float(1 + sin(clock * P.targetPulseRate) * P.targetPulse))
        case .shot:
            let goal = SIMD2(0.0, s.players[c].team == 0 ? Tuning.Pitch.goalLineZ : -Tuning.Pitch.goalLineZ)
            end = ball + simd_normalize(goal - me) * simd_distance(goal, ball)
            colour = 1
        case .unassisted:
            end = ball + A.freeLength * SIMD2(sin(angle), cos(angle))
        }
        if colour != aimKind {
            aimKind = colour
            for e in [aim, arrow] { e.model?.materials = [aimColours[colour]] }
        }
        let v = end - ball
        let length = Float(simd_length(v))
        let head = min(Float(A.width) * 3, length * 0.5)
        let yaw = simd_quatf(angle: Float(atan2(v.x, v.y)), axis: [0, 1, 0])
        let start = SIMD3(Float(ball.x), 0.07, Float(ball.y))
        let dir = SIMD3(Float(v.x), 0, Float(v.y)) / max(length, 1e-4)
        aim.isEnabled = true
        aim.position = start
        aim.orientation = yaw
        aim.scale = SIMD3(Float(A.width), 1, max(length - head, 0.01))
        arrow.isEnabled = true
        arrow.position = start + dir * (length - head)
        arrow.orientation = yaw
        arrow.scale = SIMD3(Float(A.width) * 2.6, 1, head)
    }
}
